### ISSUE TITLE
for-of does not work with some DOM collections when target is ES6

### ISSUE THREAD

**Issue by Arnavion** [Microsoft/TypeScript#2695]
Update: This issue is now only for the following collections:
- MediaList
- StyleSheetList
- CSSRuleList

Originally it was about NodeList as well, but that was fixed (along with DOMTokenList) in #3393

---

Original:

``` typescript
var paragraphs = document.querySelectorAll("p");
for (let p of paragraphs) {
    console.log(p);
}
```

```
foo.ts(2,15): error TS2488: The right-hand side of a 'for...of' statement must have a '[Symbol.iterator]()' method that returns an iterator.
```

Seems it just needs an update to lib.es6.d.ts to add the `[Symbol.iterator()]` to all the collections that have an indexer and length property. ~~The ones I found that have it in Nightly are below. The rest were either IE-only or didn't exist.~~ Removed my list since it seems FF has more than the specs allow. See https://github.com/zloirock/core-js/issues/137#issuecomment-159531781 for a more accurate list.

**Comment by mhegazy** [comment]
@zhengbli would this be covered in your change?

**Comment by Arnavion** [comment]
(This would also require making two versions of dom.d.ts - for ES5 and ES6.)

**Comment by mhegazy** [comment]
yup. that is correct.

**Comment by djarekg** [comment]
This should also be supported in ES3/ES5 since the emitted loop is just looping over the indexes.

**Comment by mhegazy** [comment]
iterators relay on symbols and these are defined in ES6. I think what we need is to allow iterating `ArrayLike` in ES3/ES5 (#2862).

**Comment by Arnavion** [comment]
@djarekg That is not correct. It can also require converting NodeList to an array first (with slice) to handle the livelist getting modified during iteration. See https://github.com/Microsoft/TypeScript/issues/2696

**Comment by NekR** [comment]
Please do not forget to include in this list ```TouchList``` interface. Thanks.

**Comment by mhegazy** [comment]
@SaschaNaz has added support for:
 * NodeList
 * NodeListOf
 * DOMTokenList

The other interfaces do not seem to be iterable in the spec. @Arnavion any ideas?

**Comment by Arnavion** [comment]
They're not `iterable<>` but they *are* ArrayClass, eg [CSSRuleList](http://www.w3.org/TR/cssom/#cssrulelist). That means [their prototype is supposed to be an Array instead of Object](http://www.w3.org/TR/WebIDL/#ArrayClass) and so they have all the properties of Array including [Symbol.iterator], although no browser has actually gone *that* far, not even Nightly.

NodeList is also supposed to be ArrayClass instead of `iterable<>`, but [Chrome had problems with it and temporarily(?) abandoned it.](https://code.google.com/p/chromium/issues/detail?id=229398) FF has [an open bug](https://bugzilla.mozilla.org/show_bug.cgi?id=869376) to convert NodeList from iterable to ArrayClass but so far nothing has happened. I don't know what browsers' plan for these "ArrayClass" interfaces - I would guess it's likely that all ArrayClass interfaces will eventually be iterable even if they aren't full-blown Arrays.

The array iterator `Array.prototype[Symbol.iterator]` is also usable for all DOM collections, and this is what Nightly does. For example, `CSSRuleList.prototype[Symbol.iterator] === Array.prototype[Symbol.iterator]` is true in Nightly. This is intentional - ArrayIterator is supposed to work with ArrayLikes. So even in browsers where DOM collection prototypes don't already have a Symbol.iterator property, assigning it with, say `CSSRuleList.prototype[Symbol.iterator] = Array.prototype[Symbol.iterator];`, works and makes them iterable with for-of. This would be an argument for TS to allow for-of **for** these types. The user just needs to add a polyfill to assign the [Symbol.iterator] property on all the DOM list prototypes they're interested in iterating with for-of. But these will have to be custom polyfills - I don't know any polyfill library that adds Symbol.iterator properties to *all* DOM collections. Babel (core-js) does it but only for NodeList.

On the other hand, Array.from() works for them since they are array-likes, and even in TS they implicitly extend ArrayLike&lt;T&gt; so Array.from()'s signature is not a problem. So that's an argument **against** allowing for-of for these types in TS.

So basically, from the user's point of view, either they use a polyfill for Array.from, and use Array.from to convert all their DOM collections before using for-of. This requires no change from TS, and such a polyfill already exists. (Then there was no point to adding [Symbol.iterator] to NodeList either, but anyway...)

Or, TS allows for-of with all DOM collections, and the user uses a polyfill that adds [Symbol.iterator] properties to all those collections. This does require change from TS, and such a polyfill doesn't already exist.

I don't have an opinion either way.

**Comment by impinball** [comment]
@Arnavion

Neither [NodeList](https://dom.spec.whatwg.org/#interface-nodelist) and [DOMTokenList](https://dom.spec.whatwg.org/#interface-domtokenlist) are supposed to be Array subclasses. They are only supposed to be iterable, as in:

- JavaScript:
  ```js
  NodeList.prototype[Symbol.iterator] = function* () {
     // ...
  }
  ```

- Python:
  ```py
  class NodeList:
    # ...
    def __iter__(self):
      # ...
  ```

- Java:
  ```java
  import java.util.*;
  public class NodeList implements Iterable<Node> {
    // ...
    Iterator<Node> iterator() {
      // ...
    }
  }
  ```

- Scala:
  ```scala
  import scala.collection.{Iterable, Iterator}
  class NodeList extends Iterable[Node] {
    // ...
    def iterator: Iterator[Node] = // ...
  }
  ```

**Comment by Arnavion** [comment]
@impinball Yes they are supposed to be, per the spec. I gave links for the other CSSOM objects in the previous post, and the link for NodeList is [here](http://www.w3.org/TR/2015/WD-dom-20150618/#old-style-collections:-nodelist-and-htmlcollection) where you can also see it's described as ArrayClass.

I already explained that no browser has actually implemented them as ArrayClass.

**Comment by impinball** [comment]
@Arnavion I see what you're talking about in the edit...This bug is actually about the following problem:

```ts
interface Foo {
  *[Symbol.iterator](): Iterator<any>;
}

declare class Bar {
  *[Symbol.iterator](): Iterator<any>;
}

declare let baz: {
  *[Symbol.iterator](): Iterator<any>;
};

declare let foo: Foo;
let bar = new Bar();

for (let _ of foo) {} // Fail?
for (let _ of bar) {} // Fail?
for (let _ of baz) {} // Fail?
```

Does that repro in any of the three cases? I don't have a dev version handy to test it.

**Comment by Arnavion** [comment]
Without the `*` (asterisks are not needed for declarations), yes all of those compile, as they're supposed to. This bug is about adding Symbol.iterator properties to the DOM types in the first place.

**Comment by impinball** [comment]
Okay...that is probably trivial

**Comment by impinball** [comment]
Unless it's something else...[the declarations already exist.](https://github.com/Microsoft/TypeScript/blob/master/src/lib/dom.es6.d.ts)

**Comment by impinball** [comment]
That file was merged on [June 5](https://github.com/Microsoft/TypeScript/commit/cc858071033b9daa8ee2d664655a59b422f5776b).

**Comment by Arnavion** [comment]
Please read this thread...

**Comment by impinball** [comment]
It type-checks for me with the latest compiler with `--target es6`. I just checked on my machine.

**Comment by dashaus** [comment]
I'm using `core-js` with target to ES5 in my project, but is impossible use `for...of` with `NodeList` type.

```typescript
let nodes = document.querySelectorAll(".nodes");

for (let node of nodes) {
}
```

When I try to compile it, I get the follow error:

```
error TS2495: Type 'NodeList' is not an array type or a string type.
```

**Comment by Arnavion** [comment]
@dashaus This issue is for when the target is ES6. See #2696 for when the target is ES5.

**Comment by zloirock** [comment]
> I don't know any polyfill library that adds Symbol.iterator properties to all DOM collections. Babel (core-js) does it but only for NodeList.

`core-js` doesn't add it because of [Chromium 38-40 bug](https://github.com/zloirock/core-js/issues/37#issuecomment-75811905). We can ignore this problem in obsolete desktop browsers, but it's also in Android 5.1. Possible replace `Symbol.iterator` to string, but it's requires wrapping `Symbol` and replacement all iterators.

**Comment by impinball** [comment]
@zloirock [Does this commit actually add support](https://github.com/zloirock/core-js/commit/558a0b99cd636676f398565aac43b8886f1bce9c)?

And I know there's still an industry for Chrome apps and Firefox-specific add-ons. And the latest few versions of those don't have this bug.

**Comment by zloirock** [comment]
@impinball only for engines w/o this bug and for this reason it's not documented.

**Comment by rsms** [comment]
Just got bitten by this. Compiling for ES6 (Chromium) and are using the new collection types (Map, Set, etc) extensively. Been using Babel for a while, but the company I work recently switched to TypeScript, and while writing a new program I noticed that a part of the program compiling fine and running fine, but producing incorrect results. It contains code that looks like this:

```ts
let entries = new Set<string>
entries.add('foo')
entries.add('bar')
// ...
for (let entry of entries.values()) {
  performTask(entry)
}
// ...
if (entries.has('bar')) {
  // read result produced by performTask
}
```

It compiles without errors and runs without exceptions. Targeting ES6/ES2015, the generated code looks like this:

```ts
var entries = new Set;
entries.add('foo');
entries.add('bar');
// ...
for (var _i = 0, _a = entries.values(); _i < _a.length; _i++) {
  var entry = _a[_i];
  performTask(entry);
}
// ...
if (entries.has('bar')) {
  // read result produced by performTask
}
```

## Proposal

1. Fix the type checker for `for..of` to check the type of the subject. For ES6 collections, it's `IterableIterator<T>` (lib/lib.es6.d.ts). Until v2.1, issue an error or warning. For and beyond v2.1 and for target=es6, emit `for..of` code instead of `for;;;` code.

2. For ES6 targets where the subject of a `for..of` statement is not of the type `IterableIterator<T>`, generate code that uses `let` instead of `var`

Example of why \#2 above is important:

input
```ts
for (let entry of entries) {
}
for (let entry of entries) {
}
```

output
```js
for (var _i = 0, _a = entries; _i < _a.length; _i++) {
    var entry = _a[_i];
}
for (var _b = 0, _c = entries; _b < _c.length; _b++) {
    var entry = _c[_b];
}
```

There're two var declarations declaring the same symbol.

**Comment by SaschaNaz** [comment]
>the generated code looks like this

The result should be just ES6 but the generated one is ES5. I think your compiler setting is wrong somewhere...

**Comment by rsms** [comment]
@SaschaNaz: Oh shoot. You're right. I do have `"target": "ES2015"` in tsconfig.json but then the build script I'm using explicitly sets --target=ES5

**Comment by mhegazy** [comment]
PRs welcomed. The change should be in https://github.com/Microsoft/TypeScript/blob/master/src/lib/dom.iterable.d.ts

**Comment by SaschaNaz** [comment]
@mhegazy Please consider Microsoft/TSJS-lib-generator#235 too ;)

**Comment by KeithHenry** [comment]
This has had no activity for a year - is it still considered a problem?

The issue still appears to be extant:

```typescript
const collection: HTMLCollection = getCollection();
const arrayCopy = [...collection]; // (TS) Type must have a '[Symbol.iterator]()' method that returns an iterator.
```

**Comment by saschanaz** [comment]
@KeithHenry I've been actively fixing wrong types on https://github.com/Microsoft/TSJS-lib-generator, I think this also will be fixed within months, or you can send your own PR there.

**Comment by KeithHenry** [comment]
@saschanaz cheers, but that already has [`HTMLCollection` configured as iterable](https://github.com/Microsoft/TSJS-lib-generator/blob/cfa8a81da649cdd296588ec03a50cb038d5ccf37/baselines/dom.es6.generated.d.ts#L45-L47). What else needs to change?

**Comment by saschanaz** [comment]
That file should replace [lib.dom.iterable.d.ts](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.iterable.d.ts) but the upstream does not yet support methods including `entries()`. Coming Soon™️, though!

**Comment by saschanaz** [comment]
Ah, by upstream I mean [this one](https://github.com/Microsoft/TSJS-lib-generator/blob/master/baselines/dom.es6.generated.d.ts).

**Comment by thany** [comment]
This problem still exists for me when trying to iterate over `NodeListOf<Element>`. A NodeList is _always_ iterable. Period.

Some code:
```javascript
const triggers = element.querySelectorAll('.trigger');
for (const trigger of triggers) {
    console.log(trigger);
}
```
The error is on `triggers` in VS Code, using typescript 3.2.2.

The error:
```
[ts] Type 'NodeListOf<Element>' is not an array type or a string type. [2495]
```
Technically correct. A NodeList is indeed neither an array nor a string. But `for..of` works on a great many more kinds of objects, including NodeList.

The above code compiles and works fine. So why complain about something that plainly **isn't the case**?

**Comment by saschanaz** [comment]
@thany Ensure the target is ES6 or higher. Currently TS does not support for-of on general iterable objects when the target is ES5/ES3.

**Comment by RyanCavanaugh** [comment]
Or enable `--downlevelIteration`

**Comment by thany** [comment]
First of all, why does TS need to support it at all? It just needs to output the transpiled JS and let JS handle whether the object can be looped over or not.

Apart from that, NodeList is _always_ iterable, in every browser. Some require a indexed for-loop, newer browser suport for..of. But either way the object can be looped over perfectly fine.

If the target is ES5, produce a classic for-loop.
If the target is ES6 or higher, produce a for..of loop.

I don't see how this is complicated in any way.

**Comment by saschanaz** [comment]
@thany Have you tried --downlevelIteration (introduced in TS2.3, I forgot about it 😅)? I think that should work.

**Comment by thany** [comment]
I've enabled `downlevelIteration` which doesn't produce any difference. The code still builds and works fine, but I still get the error in VS Code.

**Comment by saschanaz** [comment]
You may file a new issue with a repro then.

________________________________
